src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketStream.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    /// <summary>Provides a <see cref="Stream"/> that delegates to a wrapped <see cref="WebSocket"/>.</summary>
+    public sealed class WebSocketStream : Stream
+    {
+        /// <summary>Whether disposing this instance should dispose <see cref="WebSocket"/>.</summary>
+        private readonly bool _ownsWebSocket;
+        /// <summary>Whether the instance has been disposed.</summary>
+        private bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WebSocketStream"/> class using a specified <see cref="WebSocket"/> instance.
+        /// </summary>
+        /// <param name="webSocket">The <see cref="WebSocket"/> wrapped by this instance.</param>
+        /// <param name="ownsWebSocket">
+        /// <see langword="true"/> to indicate that the <see cref="WebSocketStream"/> takes ownership of the <see cref="WebSocket"/>,
+        /// such that disposing of this <see cref="Stream"/> will dispose of the <see cref="WebSocket"/>; otherwise, <see langword="false"/>.
+        /// When <see langword="true"/>, disposing of this instance doesn't initiate a close handshake; it merely delegates to
+        /// <see cref="WebSocket.Dispose"/>.
+        /// </param>
+        public WebSocketStream(WebSocket webSocket, bool ownsWebSocket = false)
+        {
+            ArgumentNullException.ThrowIfNull(webSocket);
+
+            WebSocket = webSocket;
+            _ownsWebSocket = ownsWebSocket;
+        }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+            _disposed = true;
+
+            if (disposing && _ownsWebSocket)
+            {
+                if (WebSocket.State is WebSocketState.Open)
+                {
+                    // There's no synchronous close, so we're forced to do sync-over-async.
+                    WebSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None).GetAwaiter().GetResult();
```

Comment:
Wouldn't this cause deadlocks in UI contexts such as WPF and Windows Forms?

Reply:
Not as long as everything within CloseOutputAsync uses ConfigureAwait(false)